src/blocks/calendar.rs

```diff
@@ -0,0 +1,396 @@
+//! - `$url`: Url of the event
+//! - `$location`: Location of the event
+//! - `$start`: Start time of the event
+//! - `$end`: End time of the event
+//!
+//! # Icons Used
+//! - `calendar`
+
+use chrono::{Duration, Utc};
+use oauth2::{AuthUrl, ClientId, ClientSecret, TokenUrl, Scope};
+use url::Url;
+
+use crate::{subprocess::spawn_process, util::has_command};
+
+mod auth;
+mod caldav;
+
+use self::auth::{AuthorizeUrl, OAuth2Flow, TokenStore};
+use self::caldav::Event;
+
+use super::prelude::*;
+
+use std::path::Path;
+
+use caldav::CalDavClient;
+
+#[derive(Deserialize, Debug)]
+pub struct BasicAuthConfig {
+    username: Option<String>,
+    password: Option<String>,
+}
+
+#[derive(Deserialize, Debug, SmartDefault)]
+#[serde(deny_unknown_fields, default)]
+pub struct OAuth2Config {
+    pub client_id: Option<String>,
+    pub client_secret: Option<String>,
+    pub auth_url: String,
+    pub token_url: String,
+    #[default("~/.config/i3status-rust/calendar.auth_token".into())]
+    pub auth_token: ShellString,
+    #[default(8080)]
+    pub redirect_port: u16,
+    pub scopes: Vec<Scope>,
+}
+
+#[derive(Deserialize, Default, Debug)]
+#[serde(tag = "type", rename_all = "lowercase")]
+pub enum AuthConfig {
+    #[default]
+    Unauthenticated,
+    Basic(BasicAuthConfig),
+    OAuth2(OAuth2Config),
+}
+
+#[derive(Deserialize, Debug, SmartDefault)]
+#[serde(deny_unknown_fields, default)]
+pub struct Config {
+    pub next_event_format: FormatConfig,
+    pub ongoing_event_format: FormatConfig,
+    pub no_events_format: FormatConfig,
+    pub redirect_format: FormatConfig,
+    #[default(60.into())]
+    pub interval: Seconds,
+    pub url: String,
+    pub auth: AuthConfig,
+    pub calendars: Vec<String>,
+    #[default(48)]
+    pub events_within_hours: u64,
+    #[default(300.into())]
+    pub warning_threshold: Seconds,
+    #[default("xdg-open".into())]
+    pub browser_cmd: ShellString,
+}
+
+pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
+    let next_event_format = config
+        .next_event_format
+        .with_default(" $icon $start.datetime(f:'%a %H:%M') $summary ")?;
+    let ongoing_event_format = config
+        .ongoing_event_format
+        .with_default(" $icon $summary (ends at $end.datetime(f:'%H:%M')) ")?;
+    let no_events_format = config.no_events_format.with_default(" $icon ")?;
+    let redirect_format = config
+        .redirect_format
+        .with_default(" $icon Check your web browser ")?;
+
+    api.set_default_actions(&[(MouseButton::Left, None, "open_link")])?;
+
+    let mut client = caldav_client(config).await?;
+
+    let mut timer = config.interval.timer();
+
+    let mut actions = api.get_actions()?;
+
+    loop {
+        let mut widget = Widget::new().with_format(no_events_format.clone());
+        widget.set_values(map! {
+            "icon" => Value::icon("calendar"),
+        });
+
+        let mut next_event = None;
+        let mut retries = 0;
```

Comment:
I'd probably use a for loop (same thing for your other retry loop). Also I think you want to handle the case when the AskUser failed on the first attempt. Something like:

```diff
diff --git a/src/blocks/calendar.rs b/src/blocks/calendar.rs
index efb5aa114..8a6220e69 100644
--- a/src/blocks/calendar.rs
+++ b/src/blocks/calendar.rs
@@ -210,8 +210,7 @@ pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
         });
 
         let mut next_event = None;
-        let mut retries = 0;
-        while retries <= 1 {
+       for retries in 0..=1{
             let next_event_result = get_next_event(config, &mut client).await;
             match next_event_result {
                 Ok(event) => {
@@ -223,12 +222,7 @@ pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
                         let authorization =
                             client.authorize().await.error("Authorization failed")?;
                         match &authorization {
-                            auth::Authorize::Completed => {
-                                return Err(Error::new(
-                                    "Authorization failed. Check your configurations",
-                                ))
-                            }
-                            auth::Authorize::AskUser(AuthorizeUrl { url, .. }) => {
+                            auth::Authorize::AskUser(AuthorizeUrl { url, .. }) if retries == 0 => {
                                 widget.set_format(redirect_format.clone());
                                 api.set_widget(widget.clone())?;
                                 open_browser(config, url).await?;
@@ -237,11 +231,15 @@ pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
                                     .await
                                     .error("Ask user failed")?;
                             }
+                            _ => {
+                                return Err(Error::new(
+                                    "Authorization failed. Check your configurations",
+                                ))
+                            }
                         }
                     }
                 }
             };
-            retries += 1;
         }
         if let Some(event) = next_event.clone() {
             if let (Some(start_date), Some(end_date)) = (event.end_at, event.end_at) {
```